### PR TITLE
(WIP) Update cloudfront behaviors to behave like lists and allow for ordering

### PR DESCRIPTION
**Pull request by PayscaleNateW**
#7253
- Updated references to sets instead of lists
- Updated tests so they pass
- Successfully added ordered rules to cloudfront

However, some setbacks:
Still encountering a bug that maybe someone can help me with... I get an index out of bounds exception when I run `terraform apply` due to this line:

cloudfront_distribution_configuration_structure.go
`ForwardedValues:      expandForwardedValues(m["forwarded_values"].(*schema.Set).List()[0].(map[string]interface{})),
`

There may also be a bug in field_reader.go addrToSchema. I'm not sure that code is robust against crazy combinations of nested sets and lists.

TODO:
1. Fix index out of bounds exception
2. Write new tests for this behavior
3. Update documentation about the new syntax for cache_behaviors
4. If necessary, concoct some migration plan from sets to lists... Then again, seemed to maybe work fine just making the changes to the tf file? Maybe to be safe, we can keep support for the set in a deprecated way.

### CODE REVIEW

**Comment by PayscaleNateW** [comment]
Well, tempted to close this... seems like you have it under control!

File: builtin/providers/aws/cloudfront_distribution_configuration_structure.go [hunk 1]
```diff
@@ -242,12 +242,13 @@ func expandCacheBehaviors(s *schema.Set) *cloudfront.CacheBehaviors {
...
 }
 
-func flattenCacheBehaviors(cbs *cloudfront.CacheBehaviors) *schema.Set {
+func flattenCacheBehaviors(cbs *cloudfront.CacheBehaviors) []interface{} {
 	s := []interface{}{}
 	for _, v := range cbs.Items {
 		s = append(s, flattenCacheBehavior(v))
 	}
-	return schema.NewSet(cacheBehaviorHash, s)
+	//return schema.NewSet(cacheBehaviorHash, s)
```
Comment (PayscaleNateW) [inline builtin/providers/aws/cloudfront_distribution_configuration_structure.go]: Whoops should delete that commented out part

**Comment by radeksimko** [comment]
Hi @PayscaleNateW 
do you think https://github.com/hashicorp/terraform/pull/7395 would fix this bug?

**Comment by PayscaleNateW** [comment]
@radeksimko Actually, #7485 fixed it perfectly, thanks! I pulled down your commits. I had a few merge conflicts with #7395 so I didn't give too much effort to cherry-pick that one.  But it looks like your fix is way more thorough. I definitely recommend merging those in

**Comment by PayscaleNateW** [comment]
Now trying to write acceptance tests  and I'm getting `InvalidArgument: The S3 bucket that you specified for CloudFront logs doesn't exist: mylogs.6205123422105706112.s3.amazonaws.com`

It doesn't seem like the logs bucket is defined anywhere in the mock configs? Also, when I do define one, there may be a race condition where the bucket doesn't get created in time for cloudfront to get it. Maybe Amazon just started doing that validity check?

I also notice that simply trying terraform apply using that as the config file has the same error. The s3 buckets get created, but not in time for cloudfront to use them.

**Comment by PayscaleNateW** [comment]
I believe there is a bug there pertaining to the logging bucket not being previously defined. However, I was running into issues with my naming when I tried to fix it. The name of the bucket should be a name, for example mylogs.<randint>, and then the logging_config should have its bucket as mylogs.<randint>.s3.amazonaws.com

I think I fixed this in
#9175

**Comment by adamb0mb** [comment]
I'd love to see this or #7395 merged!

**Comment by SpaceJudas** [comment]
Echo @adamb0mb!

**Comment by apparentlymart** [comment]
Hello @PayscaleNateW, and thanks for working on this!

As part of the the Terraform 0.10 release earlier this year, all of the Terraform providers were moved to their own repositories in [the terraform-providers GitHub organization](https://github.com/terraform-providers), and removed from the Terraform Core repository.

Unfortunately due to the fact that new issues and pull requests are being opened constantly, it was not possible for the various provider maintainers to merge all outstanding pull requests before this split, and there is no automatic way to migrate a pull request to a new repository.

As a result, this pull request can sadly no longer be applied as-is, and so I'm going to close it.

If you or someone else has the time and motivation to apply same changes to [the `aws` provider repository](https://github.com/terraform-providers/terraform-provider-aws) and open a new PR there, the maintainers of that provider should be able to review and merge it.

Thanks again for working on this, and sorry it was not able to be merged before the provider repository changes.

**Pull request closed without merging**